cardano-db-sync/src/Cardano/DbSync/Cache/FIFO.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Cardano.DbSync.Cache.FIFO (
+  FIFOCache (..),
+  empty,
+  insert,
+  lookup,
+  getSize,
+  getCapacity,
+  cleanupCache,
+) where
+
+import Cardano.Prelude hiding (empty)
+import qualified Data.Map.Strict as Map
+
+{-
+FIFOCache: A First-Ins-First-Outs cache with fixed capacity, using Map.Strict
+for key-value storage. It keeps 3 rolling Maps and only the first one is inserted.
+When the first one is full, the last Map is removed.
+
+Key operations and their complexities:
+- Insertion: O(log mapCapacity)
+- Lookup: O(log (mapCapacity * n))
+- Removal of oldest elements: O(1)
+- Size query: O(1)
+-}
+data FIFOCache k v = FIFOCache
```

Comment:
this isn't technically a FIFO cache anymore, did you find this to be more performant as I don't recall discussing this?

Reply:
This is what was initially discussed (I used 3 Maps instead of 2 to smoothen the flactuations of total Cache size)
It is some form of FIFO except it doesn't evicts a single element, but 1/3 of them when needed.

I plan to run a full sync on this pr today to check performance.

Reply:
> lookups would 0(log n) * 3

Worst case yes, but we know that the first Map has a way higher probability to hit than the other 2. So for most queries it will simply lookup a smaller Map. Also the previous implementation with the Queue, requires 2 Map operations per insert, one lookup and one delete and on a bigger Map.

Reply:
Cool, guess the stats will tell all 😅 I'm sure there's actually not much in it, I didn't keep the stats but when I tried a 2 Map approach it wasn't as good as the HashMap and the StrictMap was a little less performant than the HashMap. 🤷